src/erc3525/module.cairo

```diff
@@ -0,0 +1,363 @@
+        _slots: LegacyMap::<u256, u256>,
+        _slot_uri: LegacyMap::<u256, felt252>,
+        _contract_uri: felt252,
+        _total_minted: u256,
+        _total_value: LegacyMap::<u256, u256>,
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        TransferValue: TransferValue,
+        ApprovalValue: ApprovalValue,
+        SlotChanged: SlotChanged,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct TransferValue {
+        from_token_id: u256,
+        to_token_id: u256,
+        value: u256,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct ApprovalValue {
+        token_id: u256,
+        operator: ContractAddress,
+        value: u256
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct SlotChanged {
+        token_id: u256,
+        old_slot: u256,
+        new_slot: u256,
+    }
+
+    #[external(v0)]
+    impl ERC3525Impl of IERC3525<ContractState>{
+
+        fn value_decimals(self: @ContractState) -> u8 {
+            // [Compute] Value decimals
+            self._value_decimals.read()
+        }
+
+        fn balance_of(self: @ContractState, token_id: u256) -> u256 {
+            // [Check] Token exists
+            self._assert_minted(token_id);
+
+            // [Compute] Token value
+            self._values.read(token_id)
+        }
+
+        fn slot_of(self: @ContractState, token_id: u256) -> u256 {
+            // [Check] Token exists
+            self._assert_minted(token_id);
+            self._slots.read(token_id)
+        }
+
+        fn approve_value(ref self: ContractState, token_id: u256, operator: ContractAddress, value: u256) {
+            // [Check] Caller and operator are not null addresses
+            let caller = get_caller_address();
+            assert(!caller.is_zero(), 'ERC3525: invalid caller');
+            assert(!operator.is_zero(), 'ERC3525: invalid operator');
+
+            // [Check] Operator is not owner and caller is approved or owner
+            let unsafe_state = ERC721::unsafe_new_contract_state();
+            let owner = ERC721::ERC721Impl::owner_of(@unsafe_state, token_id);
+            assert(owner != operator, 'ERC3525: approval to owner');
+            assert(ERC721::InternalImpl::_is_approved_or_owner(@unsafe_state, caller, token_id), 'ERC3525: caller not allowed');
+            
+            // [Effect] Store approved value
+            self._approve_value(token_id, operator, value);
+        }
+
+        fn allowance(self: @ContractState, token_id: u256, operator: ContractAddress) -> u256 {
+            // [Check] 
+            let unsafe_state = ERC721::unsafe_new_contract_state();
+            let owner = ERC721::ERC721Impl::owner_of(@unsafe_state, token_id);
+            self._approved_values.read((owner, token_id, operator))
+        }
+
+        fn transfer_value_from(ref self: ContractState, from_token_id: u256, to_token_id: u256, to: ContractAddress, value: u256) -> u256 {
```

Comment:
Actually, maybe we need to revise this functin to follow the 3525 spec?

- We used transferValueFrom in Cairo Zero because we couldn't define multiple transferFrom functions (with different arguments)
- We used `to` AND `to_token_id` for the same reason.

Maybe now we can have multiple transferFrom with different arguments and follow [the spec](https://eips.ethereum.org/EIPS/eip-3525#specification)? (not sure about selectors though)

Reply:
I don't think it is possible, we will still have a clash even if the functions are in two seperate implementations of a single contract.
![image](https://github.com/carbonable-labs/cairo-erc-3525/assets/97087040/877e4db9-a184-4cdb-8b36-9e11fc15f413)
![image](https://github.com/carbonable-labs/cairo-erc-3525/assets/97087040/9b30c79a-0927-48ed-afbe-46a7f967c02c)

We could try to have:
- `transfer_value_from` and `transfer_value_from_`
- `transfer_value_from` and `transferValueFrom`
But all of this is more confusing and we chosed to not do this even in Cairo 0 so I think we could keep it like this